feg/gateway/services/n7_n40_proxy/notify/notify_handler.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2021 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package notify
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/labstack/echo/v4"
+
+	"magma/feg/gateway/sbi"
+	n7_client "magma/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl"
+	"magma/feg/gateway/services/n7_n40_proxy/n7"
+	"magma/feg/gateway/services/session_proxy/relay"
+	"magma/gateway/service_registry"
+	"magma/lte/cloud/go/protos"
+)
+
+// notify_handler implements the following
+// - Creates a HTTP server to receive notifications from PCF
+// - Handles SmPolicyUpdateNotification from PCF convert the message to proto and sends it as
+//   PolicyReAuth to SessionProxyResponder(feg_relay)
+// - Handles SmPolicyTerminateNotification from PCF convert the message to prot and sends it as
+//   AbortSession to feg_relay.
+
+const (
+	UnableToDeliver  = 3002
+	EncodedSessionId = "encodedSessionId"
+	// Shutdown timeout is used as a timeout to exit the Echo Server Shutdown
+	ShutdownTimeout = 10 * time.Second
+)
+
+type NotificationHandler struct {
+	config        *n7.N7ClientConfig
+	notifyServer  *echo.Echo
+	cloudRegistry service_registry.GatewayRegistry
+	ListenAddr    string
+}
+
+func NewStartedNotificationHandlerWithHandlers(
+	config *n7.N7ClientConfig,
+	cloudReg service_registry.GatewayRegistry,
+) (*NotificationHandler, error) {
+	handler := &NotificationHandler{
+		config:        config,
+		notifyServer:  echo.New(),
+		cloudRegistry: cloudReg,
+	}
+	err := handler.registerHandlers()
+	if err != nil {
+		return nil, fmt.Errorf("error registering handlers: %s", err)
+	}
+	err = handler.start()
+	if err != nil {
+		return nil, fmt.Errorf("error starting notification handler: %s", err)
+	}
+	return handler, err
+}
+
+// RegisterHandlers registers the UpdateNotification and TerminateNotification handlers.
+// The notification url is of the form {notifyRoot}/{encodedSessionId}/{operation}
+//   Example: http://magma-feg.magma.com/sm-policy-control/v1/MTIzNDU2Nzg5MDsxMjM0NQo=/update
+//   Here notifyRoot = http://magma-feg.magma.com/sm-policy-control/v1
+//        encodedSessionId = MTIzNDU2Nzg5MDsxMjM0NQo= (Session-Id is urlencoded)
+//        operation = update
+// This notification url is send to PCF in the SmPolicyCreate request
```

Comment:
Thanks for the description.
Lets put this 
```
// RegisterHandlers registers the UpdateNotification and TerminateNotification handlers.
// The notification url is of the form 
//	{notifyRoot}/{encodedSessionId}/{operation}
// Example: 
//   	http://magma-feg.magma.com/sm-policy-control/v1/MTIzNDU2Nzg5MDsxMjM0NQo=/update
// where 
//	notifyRoot = http://magma-feg.magma.com/sm-policy-control/v1
//      encodedSessionId = MTIzNDU2Nzg5MDsxMjM0NQo= (urlencoded Session-Id)
//      operation = update
// This notification url is send to PCF in the SmPolicyCreate request
```

Reply:
what does this mean
```
// This notification url is send to PCF in the SmPolicyCreate request
```

Is't the notification sent by the PCF to n7_proxy?

Reply:
Can you update the comment using the format provided please. It will make it more readable

Reply:
Done